tchannel/tornado/peer.py

```diff
@@ -621,30 +609,63 @@ def remove(self, hostport):
         """
         assert hostport, "hostport is required"
         peer = self._peers.pop(hostport, None)
-        if peer:
+        peer_in_heap = peer and peer.index != -1
+        if peer_in_heap:
             self.peer_heap.remove_peer(peer)
         return peer
 
+    def get(self, hostport):
+        """Get a Peer for the given destination.
+
+        A new Peer is added and returned if one does not already exist for the
+        given host-port. Otherwise, the existing Peer is returned.
+        """
+        assert hostport, "hostport is required"
+        if hostport not in self._peers:
+            self.add(hostport)
+
+        return self._peers[hostport]
+
     def add(self, peer):
-        """Add an existing Peer to this group.
+        """
+        Add an existing Peer to this group and the corresponding peer heap
+
+        if a peer instance is passed in, assert that it's not already in the
+        list then add it to the list and heap
+
+        if the hostport is in the peer list, make sure it's added to the heap
 
-        A peer for the given host-port must not already exist in the group.
+        otherwise create a new peer and add it to the list & heap
         """
         assert peer, "peer is required"
 
-        if isinstance(peer, basestring):
-            # Assume strings are host-ports
-            peer = self.peer_class(
-                tchannel=self.tchannel,
-                hostport=peer,
-                on_conn_change=self._update_heap,
+        if not isinstance(peer, basestring):
+            # Assume non-strings are Peers
+            assert peer.hostport not in self._peers, (
```

Comment:
If we take the one that already exists would that re-use the existing connection? I think we want that.